File: app/src/main/java/com/yuki/yukihub/launcherbridge/LauncherGameHubShortcutBridge.java
package com.yuki.yukihub.launcherbridge;

import android.content.pm.PackageManager;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import rikka.shizuku.Shizuku;

/** Reads GameHub desktop shortcut metadata through the user-authorized Shizuku service. */
public final class LauncherGameHubShortcutBridge {
    private LauncherGameHubShortcutBridge() {
    }

    public static boolean isShizukuRunning() {
        try {
            return Shizuku.pingBinder();
        } catch (Throwable ignored) {
            return false;
        }
    }

    public static boolean hasShizukuPermission() {
        try {
            return Shizuku.checkSelfPermission() == PackageManager.PERMISSION_GRANTED;
        } catch (Throwable ignored) {
            return false;
        }
    }

    public static void requestShizukuPermission(int requestCode) {
        Shizuku.requestPermission(requestCode);
    }

    /**
     * Returns GameHub shortcuts containing a launchable localGameId. This only reads
     * shortcut metadata; it does not modify GameHub or execute any game command.
     */
    public static List<Shortcut> loadShortcuts() throws Exception {
        if (!isShizukuRunning() || !hasShizukuPermission()) return new ArrayList<>();
        String output = runShizukuCommand(
                "uid=$(am get-current-user 2>/dev/null); case \"$uid\" in ''|*[!0-9]*) uid=0;; esac; "
                        + "for u in $uid 0; do "
                        + "cmd shortcut get-shortcuts --user $u --flags 31 com.xiaoji.egggamz 2>&1; "
                        + "cmd shortcut get-shortcuts --user $u --flags 31 com.xiaoji.egggame 2>&1; "
                        + "done; dumpsys shortcut 2>&1 | grep -i -A 40 -B 12 'com.xiaoji.egggamz\\|com.xiaoji.egggame\\|localGameId\\|local_\\|steamAppId' 2>&1");

        List<Shortcut> items = new ArrayList<>();
        HashSet<String> seen = new HashSet<>();
        String text = output == null ? "" : output.replace('\0', ' ');
        for (String line : text.split("\\r?\\n")) addIfValid(parseShortcut(line), items, seen);
        Matcher matcher = Pattern.compile("(?i)(localGameId\\s*[:=]\\s*([^,}\\]\\s]+)|\\blocal_[0-9a-f-]{8,}\\b|steamAppI[dD]\\s*[:=]\\s*[^0-9]*([0-9]+))").matcher(text);
        while (matcher.find()) {
            int start = Math.max(0, matcher.start() - 700);
            int end = Math.min(text.length(), matcher.end() + 1600);
            addIfValid(parseShortcut(text.substring(start, end)), items, seen);
        }
        items.sort((a, b) -> a.displayLabel.compareToIgnoreCase(b.displayLabel));
        return items;
    }

    private static Shortcut parseShortcut(String text) {
        if (text == null) return null;
        String localGameId = matchFirst(text, "(?i)\\blocalGameId\\b\\s*[:=]\\s*([^,}\\]\\s]+)");
        if (localGameId == null) localGameId = matchFirst(text, "(?i)\\b(local_[0-9a-f-]{8,})\\b");
        String steamAppId = matchFirst(text, "(?i)\\bsteamAppI[dD]\\b\\s*[:=]\\s*[^0-9]*([0-9]+)");
        String storedId = cleanValue(localGameId);
        if ((storedId == null || storedId.isEmpty()) && steamAppId != null && !steamAppId.trim().isEmpty()) {
            storedId = "steam:" + steamAppId.trim();
        }
        if (storedId == null || storedId.isEmpty()) return null;
        String appName = cleanValue(matchFirst(text, "(?i)\\blocalAppName\\b\\s*[:=]\\s*([^,}\\]\\r\\n]+)"));
        if (appName == null || appName.isEmpty()) appName = storedId;
        String label = cleanValue(matchFirst(text, "(?i)\\b(shortLabel|longLabel|title|name)\\b\\s*[:=]\\s*([^,}\\]\\r\\n]+)"));
        if (label == null || label.isEmpty()) label = appName;
        return new Shortcut(label, appName, storedId);
    }

    private static void addIfValid(Shortcut item, List<Shortcut> items, HashSet<String> seen) {
        if (item == null || item.localGameId.isEmpty()) return;
        if (seen.add(item.localGameId.toLowerCase(Locale.ROOT))) items.add(item);
    }

    private static String matchFirst(String text, String expression) {
        try {
            Matcher matcher = Pattern.compile(expression).matcher(text);
            return matcher.find() ? matcher.group(matcher.groupCount()) : null;
        } catch (Throwable ignored) {
            return null;
        }
    }

    private static String cleanValue(String value) {
        if (value == null) return null;
        String cleaned = value.trim().replaceAll("^[\\\"'=]+|[\\\"',;]+$", "");
        return cleaned.isEmpty() ? null : cleaned;
    }

    private static String runShizukuCommand(String command) throws Exception {
        Method method = Shizuku.class.getDeclaredMethod("newProcess", String[].class, String[].class, String.class);
        method.setAccessible(true);
        Process process = (Process) method.invoke(null, new Object[]{new String[]{"/system/bin/sh", "-c", command}, null, null});
        String output = readProcessStream(process.getInputStream()) + "\n" + readProcessStream(process.getErrorStream());
        try {
            process.waitFor();
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
        }
        return output;
    }

    private static String readProcessStream(InputStream input) {
        if (input == null) return "";
        try (InputStream stream = input; ByteArrayOutputStream output = new ByteArrayOutputStream()) {
            byte[] buffer = new byte[4096];
            int count;
            while ((count = stream.read(buffer)) >= 0) output.write(buffer, 0, count);
            return output.toString("UTF-8");
        } catch (Throwable ignored) {
            return "";
        }
    }

    public static final class Shortcut {
        public final String displayLabel;
        public final String localAppName;
        public final String localGameId;

        Shortcut(String displayLabel, String localAppName, String localGameId) {
            this.displayLabel = displayLabel == null ? "" : displayLabel;
            this.localAppName = localAppName == null ? "" : localAppName;
            this.localGameId = localGameId == null ? "" : localGameId;
        }
    }
}

File: app/src/main/java/com/yuki/yukihub/launcherbridge/LauncherGameHubShortcutBridge.kt
package com.yuki.yukihub.launcherbridge

import android.content.pm.PackageManager
import rikka.shizuku.Shizuku
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.util.Locale
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.regex.Pattern

/**
 * 通过用户授权的 Shizuku 服务读取 GameHub 桌面快捷方式元数据。
 */
object LauncherGameHubShortcutBridge {

    private const val MAX_RESPONSE_BYTES = 1024L * 1024L

    @JvmStatic
    fun isShizukuRunning(): Boolean = try {
        Shizuku.pingBinder()
    } catch (_: Throwable) {
        false
    }

    @JvmStatic
    fun hasShizukuPermission(): Boolean = try {
        Shizuku.checkSelfPermission() == PackageManager.PERMISSION_GRANTED
    } catch (_: Throwable) {
        false
    }

    @JvmStatic
    fun requestShizukuPermission(requestCode: Int) {
        Shizuku.requestPermission(requestCode)
    }

    /**
     * 返回包含可启动 localGameId 的 GameHub 快捷方式。仅读取快捷方式元数据；
     * 不修改 GameHub 或执行任何游戏命令。
     */
    @JvmStatic
    @Throws(Exception::class)
    fun loadShortcuts(): List<Shortcut> {
        if (!isShizukuRunning() || !hasShizukuPermission()) return emptyList()
        val output = runShizukuCommand(
            "uid=\$(am get-current-user 2>/dev/null); case \"\$uid\" in ''|*[!0-9]*) uid=0;; esac; "
                    + "for u in \$uid 0; do "
                    + "cmd shortcut get-shortcuts --user \$u --flags 31 com.xiaoji.egggamz 2>&1; "
                    + "cmd shortcut get-shortcuts --user \$u --flags 31 com.xiaoji.egggame 2>&1; "
                    + "done; dumpsys shortcut 2>&1 | grep -i -A 40 -B 12 'com.xiaoji.egggamz\\|com.xiaoji.egggame\\|localGameId\\|local_\\|steamAppId' 2>&1"
        )

        val items = ArrayList<Shortcut>()
        val seen = HashSet<String>()
        val text = (output ?: "").replace('\u0000', ' ')
        for (line in text.split(Regex("\\r?\\n"))) {
            addIfValid(parseShortcut(line), items, seen)
        }
        val matcher = Pattern.compile(
            "(?i)(localGameId\\s*[:=]\\s*([^,}\\]\\s]+)|\\blocal_[0-9a-f-]{8,}\\b|steamAppI[dD]\\s*[:=]\\s*[^0-9]*([0-9]+))"
        ).matcher(text)
        while (matcher.find()) {
            val start = maxOf(0, matcher.start() - 700)
            val end = minOf(text.length, matcher.end() + 1600)
            addIfValid(parseShortcut(text.substring(start, end)), items, seen)
        }
        items.sortWith(Comparator { a, b -> a.displayLabel.compareTo(b.displayLabel, ignoreCase = true) })
        return items
    }

    private fun parseShortcut(text: String?): Shortcut? {
        if (text == null) return null
        var localGameId = matchFirst(text, "(?i)\\blocalGameId\\b\\s*[:=]\\s*([^,}\\]\\s]+)")
        if (localGameId == null) localGameId = matchFirst(text, "(?i)\\b(local_[0-9a-f-]{8,})\\b")
        val steamAppId = matchFirst(text, "(?i)\\bsteamAppI[dD]\\b\\s*[:=]\\s*[^0-9]*([0-9]+)")
        var storedId = cleanValue(localGameId)
        if (storedId.isNullOrEmpty() && !steamAppId.isNullOrBlank()) {
            storedId = "steam:${steamAppId.trim()}"
        }
        if (storedId.isNullOrEmpty()) return null
        var appName = cleanValue(matchFirst(text, "(?i)\\blocalAppName\\b\\s*[:=]\\s*([^,}\\]\\r\\n]+)"))
        if (appName.isNullOrEmpty()) appName = storedId
        var label = cleanValue(matchFirst(text, "(?i)\\b(shortLabel|longLabel|title|name)\\b\\s*[:=]\\s*([^,}\\]\\r\\n]+)"))
        if (label.isNullOrEmpty()) label = appName
        return Shortcut(label, appName, storedId)
    }

    private fun addIfValid(item: Shortcut?, items: MutableList<Shortcut>, seen: HashSet<String>) {
        if (item == null || item.localGameId.isEmpty()) return
        if (seen.add(item.localGameId.lowercase(Locale.ROOT))) items.add(item)
    }

    private fun matchFirst(text: String, expression: String): String? = try {
        val matcher = Pattern.compile(expression).matcher(text)
        if (matcher.find()) matcher.group(matcher.groupCount()) else null
    } catch (_: Throwable) {
        null
    }

    private fun cleanValue(value: String?): String? {
        if (value == null) return null
        val cleaned = value.trim().replace(Regex("^[\\\"'=]+|[\\\"',;]+$"), "")
        return cleaned.ifEmpty { null }
    }

    @Throws(Exception::class)
    private fun runShizukuCommand(command: String): String {
        val method = Shizuku::class.java.getDeclaredMethod("newProcess", Array<String>::class.java, Array<String>::class.java, String::class.java)
        method.isAccessible = true
        val process = method.invoke(null, arrayOf("/system/bin/sh", "-c", command), null, null) as Process
        val readers = Executors.newFixedThreadPool(2)
        try {
            val stdout = readers.submit<String> { readProcessStream(process.inputStream) }
            val stderr = readers.submit<String> { readProcessStream(process.errorStream) }
            if (!process.waitFor(10, TimeUnit.SECONDS)) {
                process.destroyForcibly()
                throw java.io.IOException("Shizuku command timed out")
            }
            return stdout.get(2, TimeUnit.SECONDS) + "\n" + stderr.get(2, TimeUnit.SECONDS)
        } finally {
            readers.shutdownNow()
            process.destroy()
        }
    }

    private fun readProcessStream(input: InputStream?): String {
        if (input == null) return ""
        return try {
            input.use { stream ->
                val output = ByteArrayOutputStream()
                val buffer = ByteArray(4096)
                var count: Int
                while (stream.read(buffer).also { count = it } >= 0) {
                    if (output.size() + count > MAX_RESPONSE_BYTES) throw java.io.IOException("command output too large")
                    output.write(buffer, 0, count)
                }
                output.toString("UTF-8")
            }
        } catch (_: Throwable) {
            ""
        }
    }

    class Shortcut(
        displayLabel: String?,
        localAppName: String?,
        localGameId: String?
    ) {
        @JvmField val displayLabel: String = displayLabel ?: ""
        @JvmField val localAppName: String = localAppName ?: ""
        @JvmField val localGameId: String = localGameId ?: ""
    }
}
